qa_eval/steps/sparql.py

```diff
@@ -33,11 +33,114 @@ def get_permutation_indices(list1: list, list2: list) -> list:
+
+    return True
+
+
+def compare_unordered_results(
+    reference_values: list[str],
+    actual_values: list[str],
+    permutation: list[int],
+) -> tuple[bool, list[int]]:
+    if not permutation:
+        permutation_indices = get_permutation_indices(reference_values, actual_values)
+        if permutation_indices:
+            return True, permutation_indices
+    if permutation:
+        if is_permutation(actual_values, reference_values, permutation):
+            return True, permutation
+    return False, permutation
+
+
+def compare_columns(
+    reference_values: list[str],
+    actual_values: list,
+    results_are_ordered: bool,
+    permutation: list[int],
+) -> tuple[bool, list[int]]:
+    """Compares a list of actual values against a reference list.
+
+    This function supports both order-sensitive and order-insensitive
+    comparisons. For an unordered comparison, it can either find a new
+    permutation or verify a pre-existing one.
+
+    Args:
+        reference_values: The list of expected values.
+        actual_values: The list of values to be checked.
+        results_are_ordered: If True, performs a direct element-wise
+            comparison. If False, checks if `actual_values` is a
+            permutation of `reference_values`.
+        permutation: An optional list of indices. For unordered checks,
+            if provided, it's used to verify a specific permutation.
+
+    Returns:
+        A tuple containing:
+            - bool: True if the lists match (either directly or as a
+              permutation)
+            - list[int]: The list of permutation indices. If an unordered
+              match is found, ordering is not required, this list maps the indices of
+              `reference_values` to `actual_values`.
+    """
+    if not results_are_ordered:
+        if not permutation:
+            permutation_indices = get_permutation_indices(
+                reference_values, actual_values
+            )
+            if permutation_indices:
+                return True, permutation_indices
+        if permutation:
+            if is_permutation(actual_values, reference_values, permutation):
+                return True, permutation
+        return False, permutation
+    if reference_values == actual_values:
+        return True, permutation
+    return False, permutation
+
+
+def compare_values(
+    reference_vars: list[str],
+    reference_var_to_values: dict[str, list],
+    actual_vars: list[str],
+    actual_var_to_values: dict[str, list],
+    required_vars: list[str],
+    results_are_ordered: bool,
+) -> bool:
+    permutation = []
+    mapped_or_skipped_reference_vars, mapped_actual_vars = set(), set()
+    for reference_var in reference_vars:
+        reference_values = reference_var_to_values[reference_var]
+        for actual_var in actual_vars:
+            if actual_var in mapped_actual_vars:
+                continue
+            new_mapping, permutation = compare_columns(
```

Comment:
The variable name `new_mapping` is confusing. I am not completely sure how it is used, but because it is a boolean variable, a better name may be `is_new_mapping`.